core/src/main/java/com/datastax/oss/driver/api/core/data/CqlVector.java

```diff
@@ -15,25 +15,31 @@
  */
 package com.datastax.oss.driver.api.core.data;
 
-import com.datastax.oss.driver.shaded.guava.common.base.Joiner;
-import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
-import com.datastax.oss.driver.shaded.guava.common.collect.Iterators;
-import java.util.Arrays;
+import java.util.*;
 
-/** An n-dimensional vector defined in CQL */
-public class CqlVector<T> {
+/**
+ * An n-dimensional vector defined in CQL. You can use either {@link CqlVector#builder()} for an
+ * iterable interface or {@link #CqlVector(List)} directly. Once created, a CqlVector is immutable.
+ */
+public class CqlVector<T> implements Iterable<T> {
 
-  private final ImmutableList<T> values;
+  private final List<T> values;
 
-  private CqlVector(ImmutableList<T> values) {
-    this.values = values;
+  /**
+   * Create a CqlVector from a list of values.
+   *
+   * @param values
+   */
+  public CqlVector(List<T> values) {
+    this.values = Collections.unmodifiableList(values);
   }
 
-  public static Builder builder() {
-    return new Builder();
+  public static <T> CqlVector of(T... values) {
+    return new CqlVector(Collections.unmodifiableList(Arrays.asList(values)));
   }
 
-  public Iterable<T> getValues() {
+  /** @return the (immutable) list of values in the vector */
+  public List<T> getValues() {
     return values;
   }
```

Comment:
Upon reflection I'm not sure we need a method to just return the contained values here.  Supporting Iterable should cover the overwhelming majority of cases; my inclination is to just go with that until we see we need something else.

Reply:
This, I emphatically disagree with, as has any other developer which I've seen try to use it or discuss it so far. This is the most important aspect of the change we need. Here is a scenario to illustrate:

- We need to run an intensive performance test that is logically and operationally repeatable.
- We need to run a variation of the performance test which has the simplest possible change, from un-normalized vectors to normalized vectors. The way to do that is, arguably, to simply allow for a CqlVector to be normalized. 
- We provide a basic function to do that ^, but because we are forced to use an iterator interface with (with no size data, even), we must:
  - create a collection on heap, which will have to be dynamically resized and reallocated each time we exceed its backing capacity, which is incredibly wasteful for larger vectors.
  - iteratively walk the data and store it into the collection
  - run our computation against the separate copy of the data to compute the normalization factor
  - run another scaling computation against those values
  - (and with the forced builder pattern) construct a whole new object, install the values into, and ask it for our new CqlVector
- Aspects of our testing mean that we're spending *significant* cycles on just creating and modifying vectors.

The very very prescriptive design of the original type might make things look easier according to one lens, but it definitely does *not* make it easier nor reasonably performant for those who want to do the following:
- use simple idiomatic interfaces to create value types (Builder is not idiomatic for *all* cases)
- use simple idiomatic interfaces to access efficient data from value types (Iterator is not idiomatic for *all* cases)
- write less code which is easy as or more self-evident than a builder pattern
- be able to identify a simple and appropriate method for constructing a value type without spelunking throug API docs
- have an efficient implementation that doesn't completely nerf an application when this type is used in any significant proportion of cycles
- think of vectors as _vectors_, as nearly every ML practitioner who directly works with vectors (that's a large fraction of the ML industry at this point) will see this interface and scratch their head as to why they are having to do these complicated things in Java just to use a _vector_.

While I do agree with the desire to be prescriptive, there is a good way to do this and a way that simply forces everyone to "do it your way". This is one of those case. Simple value type based on POJOs should not come with this degree of ceremony nor serious performance trade-offs. It's supposed to be *simple*. That's what a good value type should provide for the user.

In order to overcome the testing impact of the iterative accessor, we've had to create our own branches and build from them. We could jump through some hoops to prove out the impacts, but I thought they were evident enough that we wouldn't need to spend hours or days to justify such a direct and simple improvement.

Reply:
Understood, this explanation is helpful... thanks @jshook!

FTR in my working PR I've tried to address this by adding a method to export data from a vector to an input collection.  If I understand correctly (always a question) this will allow users to do the "could to this instead" section above while still making no assumptions about the type of collection involved or any of it's properties (other than the fact that it's, you know, a Collection).

We'll see if that flies.

Reply:
without knowing the details -- maybe?

the core issue is we need to be able to modify it w/o copying it

copies of large collections like vectors really hurt perf

Reply:
And copying them incrementally with a dynamically sized backing store, one element at a time, is a good way to test your GC, because this is a lot of small allocations.

> As elements are added to an ArrayList, its capacity grows automatically. The details of the growth policy are not specified beyond the fact that adding an element has constant amortized time cost.

And the nitty gritty details are almost always "the existing backing memory is copied into a newly allocated region", so you can see how this would create a heap allocation trampoline.

Reply:
I believe this entire discussion is now rendered moot by virtue of the most recent round of changes to my working PR for JAVA-3061.  I won't repeat the discussion in my most [recent comment on the topic](https://github.com/datastax/java-driver/pull/1656#issuecomment-1605121141) but the upshot is that CqlVector no longer exists and the Java driver now just represents CQL vectors as Java Lists.